packages/arb-avm-cpp/data_storage/src/datastorage.cpp

```diff
@@ -220,3 +260,46 @@ rocksdb::Status DataStorage::updateSecretHashSeed() {
     }
     return status;
 }
+
+ConcurrentCounter DataStorage::getCounter() const {
+    return ConcurrentCounter::Get(this);
+}
+
+ConcurrentCounter::ConcurrentCounter(const DataStorage* _storage)
+    : storage(_storage) {
+    storage->concurrent_database_access_counter++;
+    if (storage->shutting_down) {
+        // Destructor will take care of decrementing counter
+        throw DataStorage::shutting_down_exception();
+    }
+}
+
+ConcurrentCounter::~ConcurrentCounter() {
+    if (storage == nullptr) {
```

Comment:
Would be nice to have a comment saying that this makes it so the destructor of a counter that's been moved won't run

Reply:
:+1: